flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -613,25 +614,31 @@ protected void cleanUpInvoke() throws Exception {
 		// stop all timers and threads
 		tryShutdownTimerService();
 
+		Throwable suppressedThrowable = null;
```

Comment:
So we have 4-5 similar try-catch blocks one after another now.

I guess we can't use Guava `Closer` because we want to handle any type of exception, right?
And not CloseableRegistry because it closes silently.

Maybe we can emulate it by having a collection of closeables and close them in a loop:
```
List<ThrowingRunnable<?>> runs = asList(cancelables::close, this::shutdownAsyncThreads, this::cleanup, this::disposeAllOperators, ...);
Throwable suppressedThrowable = null;
for (ThrowingRunnable<?> run: runs) {
    try {
        run.run();
    } catch (Throwable t) {
        suppressedThrowable = ExceptionUtils.firstOrSuppressed(t, suppressedThrowable);
    }
}
```

Reply:
> I guess we can't use Guava `Closer` because we want to handle any type of exception, right?
> And not CloseableRegistry because it closes silently.

Hmm, need more context to understand this part.

Reply:
> Maybe we can emulate it by having a collection of closeables and close them in a loop:
> 
> ```
> List<ThrowingRunnable<?>> runs = asList(cancelables::close, this::shutdownAsyncThreads, this::cleanup, this::disposeAllOperators, ...);
> Throwable suppressedThrowable = null;
> for (ThrowingRunnable<?> run: runs) {
>     try {
>         run.run();
>     } catch (Throwable t) {
>         suppressedThrowable = ExceptionUtils.firstOrSuppressed(t, suppressedThrowable);
>     }
> }
> ```

yep, that's a good way, if we decide to suppress every exception.

I do not like the way we have so many try-catch blocks as well :-)

Reply:
> Hmm, need more context to understand this part.

You can take a look at in `StreamTask.cancelables` or usages of `org.apache.flink.shaded.guava18.com.google.common.io.Closer` (but as I said, I think they can't be used here).

Reply:
yes, that's a good idea, let's do this.